src/Pinch/Protocol/Compact.hs

```diff
@@ -0,0 +1,487 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+-- |
+-- Module      :  Pinch.Protocol.Compact
+-- Copyright   :  (c) Ben Gamari 2015
+-- License     :  BSD3
+--
+-- Maintainer  :  Abhinav Gupta <mail@abhinavg.net>
+-- Stability   :  experimental
+--
+-- Implements the Thrift Compact Protocol as a 'Protocol'.
+module Pinch.Protocol.Compact (compactProtocol) where
+
+
+#if __GLASGOW_HASKELL__ < 709
+import Control.Applicative
+#endif
+
+import Control.Monad
+import Data.List           (sortBy)
+import Data.Ord            (comparing)
+import Data.Bits           hiding (shift)
+import Data.ByteString     (ByteString)
+import Data.Foldable       (null)
```

Comment:
Also, it turns out that the null check on FoldList is really expensive because it'll have to iterate through the whole collection. Instead, we can use the sizes which we get anyway when we are serializing the objects.